### PR TITLE
rgw/s3select : fix for error flow. add an option to disable s3select-request.

### PR DESCRIPTION
**Pull request by galsalomon66**
-- add the option to reject s3select-requests 
`ceph config set client.rgw.8000 rgw_s3select_disable true`

-- in some cases the error message does not return to client, connection got broken (invalid chunk length)

https://bugzilla.redhat.com/show_bug.cgi?id=2252396
https://bugzilla.redhat.com/show_bug.cgi?id=2275323
https://github.com/ceph/s3-tests/pull/561  NOTE:  the s3-tests PR must merged with this PR

https://tracker.ceph.com/issues/65828 

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by galsalomon66** [comment]
https://pulpito.ceph.com/gsalomon-2024-04-19_04:12:34-rgw:verify-fix_error_message_flow-distro-default-smithi/
4 dead : not related to changes, all are centos.

**Comment by galsalomon66** [comment]
https://pulpito.ceph.com/gsalomon-2024-04-19_21:46:08-rgw:verify-fix_error_message_flow-distro-default-smithi/

**Comment by galsalomon66** [comment]
https://pulpito.ceph.com/gsalomon-2024-04-25_16:28:18-rgw:verify-fix_error_message_flow-distro-default-smithi/
all is green

File: src/rgw/rgw_s3select.cc [hunk 1]
```diff
@@ -230,14 +233,17 @@ void aws_response_handler::send_success_response()
...
+  m_fp_chunk_encoding();
   int buff_len = create_message(m_success_header_size);
   s->formatter->write_bin_data(sql_result.data(), buff_len);
   rgw_flush_formatter_and_reset(s, s->formatter);
 }
 
-void aws_response_handler::send_error_response_rgw_formatter(const char* error_code,
-    const char* error_message,
-    const char* resource_id)
+std::string empty_error="--";
```
[the remark this answers is not in the document]
Comment (galsalomon66) [reply src/rgw/rgw_s3select.cc:242]: I missed that. 
will fixed that to static/constexpr

File: qa/rgw/s3tests-branch.yaml [hunk 2]
```diff
@@ -1,4 +1,6 @@
 overrides:
   s3tests:
-    force-branch: ceph-master
+    force-branch: fix_non_handled_error_resonse
+    git_remote: https://github.com/galsalomon66/
```
[the remark this answers is not in the document]
Comment (galsalomon66) [reply qa/rgw/s3tests-branch.yaml]: these lines already removed  (back to ceph-master)

**Comment by yuvalif** [comment]
* please link to trackers not BZs

* i don't understand how this is fixing the OOM issue. clearly, the s3-select request was not sent by mistake, and the intent of the user was to use the feature. how will an admin know that there are users that indend to send large files up front?
* using the flag as a "kill switch" when they see RGW crashes (guessing the reason is s3-select OOM...) is, IMO, *very* wrong
* would recommend adding a conf parameter on the maximum size of a file allowed for s3-select. this does not have to be very accurate (1GB? 5GB?), and the admin settign the value will be able to reason about the decision to not allowing for using the feature on very large files as it may crash the RGW

**Comment by galsalomon66** [comment]
> * please link to trackers not BZs
> * i don't understand how this is fixing the OOM issue. clearly, the s3-select request was not sent by mistake, and the intent of the user was to use the feature. how will an admin know that there are users that indend to send large files up front?
> * using the flag as a "kill switch" when they see RGW crashes (guessing the reason is s3-select OOM...) is, IMO, _**very**_ wrong
> * would recommend adding a conf parameter on the maximum size of a file allowed for s3-select. this does not have to be very accurate (1GB? 5GB?), and the admin settign the value will be able to reason about the decision to not allowing for using the feature on very large files as it may crash the RGW

adding @mattbenjamin 

the "kill switch" could be "too dramatic" 
as for Parquet-size-limitation switch,  the OOM occurs not because of Parquet size but because of the Parquet chunk size
(the way Parquet is organized),
meaning, if there were much more row-groups (not just 6), it would "survive" in a low-end server.
i agree that the Parquet-size-limitation is more subtle.

**Comment by galsalomon66** [comment]
jenkins test make check

**Comment by galsalomon66** [comment]
jenkins test api

**Review by ivancich — APPROVED** [review]
lgtm

**Comment by cbodley** [comment]
i very much agree with @yuvalif's sentiments above

admins need a way to control rgw's total memory usage so it doesn't keep crashing with OOM. rgw will process up to rgw_max_concurrent_requests (1024) at a time. each Get/PutObject request will buffer up to rgw_get_obj_window_size/rgw_put_obj_min_window_size (16M) of data. so the admin can tune these knobs to guarantee that a server saturated with requests will still fit in available memory. we need a way to fit SELECT requests into this model too

so how is it that a single SELECT request can kill the server with OOM? what data is it that's not bounded in size? is it state from the parquet parser? is it buffered results that we haven't written back to the client yet?

that said, users have requested a way to disable s3select (like in https://tracker.ceph.com/issues/61821) so i'm glad to see this. i just think there's more work to do here

**Comment by galsalomon66** [comment]
@cbodley 

yes,
this issue should resolved with a memory management solution, not with disable-flag.

row groups divide the object, enabling skipping this basic building block.
I.e, in case the row-group does not match the statement predicate, we can skip it.
the bigger the row-group the less possible that it *does not* match the predicate(thus, no skipping).

this 12GB object contains 6 groups, it is definitely not efficient and also caused the OOM.

the solution should be by fetching batches(pre-defined size) of that row group whatever its size, and not all in one batch, as currently done. 

thus, filtering out big-size Parquet objects is not a solution.

on a side note, 
the different readers are integrated into the GetObject, each of each works differently, the concept is to stream data not to pre-load it.
CSV is the most simple reader, while JSON and Parquet are more complex.

in the case of Parquet, the SQL engine(actually, the Parquet reader) issues range-requests according to meta-data,
while in the JSON case, the SQL statement defines how much JSON data to read (it should be a row, that row has no pre-defined end-of-row)

the combination of the SQL statement and data-profile could sometimes lead to behavior that was not anticipated in advance.

**Comment by galsalomon66** [comment]
@cbodley @yuvalif 

the following PR https://github.com/ceph/s3select/pull/163 (WIP)
fix the memory usage consumption by limiting the buffer size internally

**Pull request merged**